course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/CourseEnrolmentActor.scala

```diff
@@ -692,6 +694,63 @@ class CourseEnrolmentActor @Inject()(@Named("course-batch-notification-actor") c
+            ProjectCommonException.throwClientErrorException(ResponseCode.accessDeniedToEnrolOrUnenrolCourse, programId);
+
+        if (!isAdminAPI && (contentData.size() == 0 || !util.Arrays.asList(getConfigValue(JsonKey.PROGRAM_ENROLL_ALLOWED_PRIMARY_CATEGORY).split(","): _*).contains(contentData.get(JsonKey.PRIMARYCATEGORY).asInstanceOf[String])))
+            ProjectCommonException.throwClientErrorException(ResponseCode.accessDeniedToEnrolOrUnenrolCourse, programId);
+
+        val userIds = request.get(JsonKey.USERID_LIST).asInstanceOf[java.util.List[String]]
+        val batchId: String = request.get(JsonKey.BATCH_ID).asInstanceOf[String]
+        val batchData: CourseBatch = courseBatchDao.readById(programId, batchId, request.getRequestContext)
+
+        for (userId <- userIds) {
+            try {
+                val enrolmentData: UserCourses = userCoursesDao.read(request.getRequestContext, userId, programId, batchId)
+                val batchUserData: BatchUser = batchUserDao.read(request.getRequestContext, batchId, userId)
+                validateEnrolment(batchData, enrolmentData, true)
+                getCoursesForProgramAndEnrol(request, programId, userId, batchId)
+                val dataBatch: util.Map[String, AnyRef] = createBatchUserMapping(batchId, userId, batchUserData)
+                val data: java.util.Map[String, AnyRef] = createUserEnrolmentMap(userId, programId, batchId, enrolmentData, request.getContext.getOrDefault(JsonKey.REQUEST_ID, "").asInstanceOf[String])
+                upsertEnrollment(userId, programId, batchId, data, dataBatch, (null == enrolmentData), request.getRequestContext)
+                logger.info(request.getRequestContext, "ProgramEnrolmentActor :: enroll :: Deleting redis for key " + getCacheKey(userId))
+                cacheUtil.delete(getCacheKey(userId))
+                generatePreProcessorKafkaEvent(request, batchId, programId, userId)
+                generateTelemetryAudit(userId, programId, batchId, data, "enrol", JsonKey.CREATE, request.getContext)
+                notifyUser(userId, batchData, JsonKey.ADD)
+                status.put(JsonKey.STATUS, JsonKey.SUCCESS)
+                response.put(userId, status)
+            } catch {
+                case e: ProjectCommonException =>
+                    if (ResponseCode.userAlreadyEnrolledCourse.getErrorMessage.equals(e.getMessage)) {
+                        map.put(JsonKey.STATUS, JsonKey.FAILED)
+                        map.put(JsonKey.ERRORMSG, ResponseCode.userAlreadyEnrolledCourse.getErrorMessage)
+                        response.put(userId, map)
+                    } else {
+                        map.put(JsonKey.STATUS, JsonKey.FAILED)
+                        map.put(JsonKey.ERRORMSG, "user not enrolled")
```

Comment:
errMsg -- should explain what is the actual error... get the information from exception and use the same.

Reply:
resolved